// cypress-accessibility-checker/src/lib/common/report/ACReporterCSV.js
"use strict";
/******************************************************************************
     Copyright:: 2020- IBM, Inc

    Licensed under the Apache License, Version 2.0 (the "License");
    you may not use this file except in compliance with the License.
    You may obtain a copy of the License at

    http://www.apache.org/licenses/LICENSE-2.0

    Unless required by applicable law or agreed to in writing, software
    distributed under the License is distributed on an "AS IS" BASIS,
    WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
    See the License for the specific language governing permissions and
    limitations under the License.
  *****************************************************************************/
var __awaiter = (this && this.__awaiter) || function (thisArg, _arguments, P, generator) {
    function adopt(value) { return value instanceof P ? value : new P(function (resolve) { resolve(value); }); }
    return new (P || (P = Promise))(function (resolve, reject) {
        function fulfilled(value) { try { step(generator.next(value)); } catch (e) { reject(e); } }
        function rejected(value) { try { step(generator["throw"](value)); } catch (e) { reject(e); } }
        function step(result) { result.done ? resolve(result.value) : adopt(result.value).then(fulfilled, rejected); }
        step((generator = generator.apply(thisArg, _arguments || [])).next());
    });
};
Object.defineProperty(exports, "__esModule", { value: true });
exports.ACReporterCSV = void 0;
const ReporterManager_1 = require("./ReporterManager");
class ACReporterCSV {
    name() {
        return "csv";
    }
    generateReport(_reportData) {
    }
    generateSummary(config, _rulesets, endReport, compressedReports) {
        return __awaiter(this, void 0, void 0, function* () {
            let toCSV = ACReporterCSV.toCSV;
            let resultStr = `Label,Level,RuleId,Message,Xpath,Help\n`;
            let startScan = 0;
            for (const compressedReport of compressedReports) {
                let reportStored = ReporterManager_1.ReporterManager.uncompressReport(compressedReport);
                if (startScan === 0) {
                    startScan = reportStored.engineReport.summary.startScan;
                }
                let report = reportStored.engineReport;
                for (const result of report.results) {
                    resultStr += `${toCSV(reportStored.label)},${toCSV(result.level)},${toCSV(result.ruleId)},${toCSV(result.message)},${toCSV(result.path.dom)},${toCSV(result.help)}\n`;
                }
            }
            let startScanD = new Date(startScan);
            let reportFilename = `results_${startScanD.toISOString().replace(/:/g, "-")}.csv`;
            if (config.outputFilenameTimestamp === false) {
                reportFilename = `results.csv`;
            }
            return {
                summaryPath: reportFilename,
                summary: resultStr
            };
        });
    }
}
exports.ACReporterCSV = ACReporterCSV;
ACReporterCSV.toCSV = function (str) {
    if (str === null) {
        return '"null"';
    }
    else if (!str || str.length == 0) {
        return '""';
    }
    else {
        str = str.replace(/"/g, '""');
        return `"${str}"`;
    }
};
